src/ioabc.cpp

```diff
@@ -304,44 +305,100 @@ void ABCInput::CalcUnitNoteLength()
     }
 }
 
-void ABCInput::AddBeam()
+void ABCInput::AddLayerElement()
 {
-    if (!m_noteStack.size()) {
+    // exit if there is nothing to add
+    if (!m_noteStack.size()) return;
+    // if just one note in the stack - add it do the layer directly
+    if (m_noteStack.size() == 1) {
+        if (m_containerElement.m_element && (ElementType::Tuplet == m_containerElement.m_type)) {
+            m_containerElement.m_element->AddChild(m_noteStack.back());
+            if (!--m_containerElement.m_count) {
+                m_layer->AddChild(m_containerElement.m_element);
+                m_containerElement = {};
+            }
+        }
+        else {
+            m_layer->AddChild(m_noteStack.back());
+        }
+        m_noteStack.clear();
         return;
     }
-    else if (m_noteStack.size() == 1) {
-        m_layer->AddChild(m_noteStack.back());
+    // otherwise we can have beam or tuplet (for now)
+    Beam *beam = new Beam();
+    // add stacked notes to the current element
+    for (auto iter = m_noteStack.begin(); iter != m_noteStack.end(); ++iter) {
+        beam->AddChild(*iter);
     }
-    else {
-        Beam *beam = new Beam();
-        for (auto iter = m_noteStack.begin(); iter != m_noteStack.end(); ++iter) {
-            beam->AddChild(*iter);
-        }
-        if (beam->FindDescendantByType(NOTE)) {
-            m_layer->AddChild(beam);
+    if (beam->FindDescendantByType(NOTE)) {
+        LayerElement *element = NULL;
+        if (m_containerElement.m_element && (ElementType::Tuplet == m_containerElement.m_type)) {
+            element = m_containerElement.m_element;
+            element->AddChild(beam);
         }
+        // otherwise default to it being beam
         else {
-            for (auto iter = m_noteStack.begin(); iter != m_noteStack.end(); ++iter) {
-                m_layer->AddChild(*iter);
-            }
+            element = beam;
+        }
+        m_layer->AddChild(element);
+    }
+    else {
+        for (auto iter = m_noteStack.begin(); iter != m_noteStack.end(); ++iter) {
+            m_layer->AddChild(*iter);
```

Comment:
Could you double check that we don't have a memory leak if beam has no descendant NOTE? I think if we end up here beam is never used or deleted